torch/distributions/multivariate_normal.py

```diff
@@ -0,0 +1,154 @@
+    :math:`\mathbf{\Sigma} = \mathbf{L}\mathbf{L}^\top` as obtained via e.g. 
+    Cholesky decomposition of the covariance.
+
+    Example:
+
+        >>> m = MultivariateNormal(torch.zeros(2), torch.eye(2))
+        >>> m.sample()  # normally distributed with mean=`[0,0]` and covariance_matrix=`I`
+        -0.2102
+        -0.5429
+        [torch.FloatTensor of size 2]
+
+    Args:
+        loc (Tensor or Variable): mean of the distribution
+        covariance_matrix (Tensor or Variable): positive-definite covariance matrix
+        scale_tril (Tensor or Variable): lower-triangular factor of covariance
+        
+    Note:
+        Only one of `covariance_matrix` or `scale_tril` can be specified.
+        
+    """
+    params = {'loc': constraints.real_vector,
+              'covariance_matrix': constraints.positive_definite,
+              'scale_tril': constraints.lower_cholesky }
+    support = constraints.real
+    has_rsample = True
+
+    def __init__(self, loc, covariance_matrix=None, scale_tril=None):
+        event_shape = torch.Size(loc.shape[-1:])
+        if (covariance_matrix is None) == (scale_tril is None):
+            raise ValueError("Exactly one of covariance_matrix or scale_tril may be specified (but not both).")
+        if scale_tril is None:
+            if covariance_matrix.dim() < 2:
+                raise ValueError("covariance_matrix must be two-dimensional")
+            self.covariance_matrix = covariance_matrix
+            batch_shape = _get_batch_shape(covariance_matrix, loc)
+        else:
+            if scale_tril.dim() < 2:
+                raise ValueError("scale_tril matrix must be two-dimensional")
+            self.scale_tril = scale_tril
+            batch_shape = _get_batch_shape(scale_tril, loc)
+        self.loc = loc
+        super(MultivariateNormal, self).__init__(batch_shape, event_shape)
+
+    @lazy_property
+    def scale_tril(self):
+        return _batch_potrf_lower(self.covariance_matrix)
+
+    @lazy_property
+    def covariance_matrix(self):
+        # Note: To use torch.bmm, we need to first squash the batch_shape into a single dimension
+        flat_scale_tril = self.scale_tril.unsqueeze(0).contiguous().view((-1,)+self._event_shape*2)
+        return torch.bmm(flat_scale_tril, flat_scale_tril.transpose(-1,-2)).view(self.scale_tril.shape)
+
+    def rsample(self, sample_shape=torch.Size()):
+        shape = self._extended_shape(sample_shape)
+        eps = self.loc.new(*shape).normal_()
+        return self.loc + _batch_mv(self.scale_tril, eps)
+
+    def log_prob(self, value):
+        self._validate_log_prob_arg(value)
+        delta = value - self.loc
+        M = _batch_mahalanobis(self.scale_tril, delta)
+        log_det = _batch_diag(self.scale_tril).abs().log().sum(-1)
+        return -0.5*(M + self.loc.size(-1)*math.log(2*math.pi)) - log_det
+
+    def entropy(self):
+        log_det = _batch_diag(self.scale_tril).abs().log().sum(-1)
```

Comment:
Hmm, shouldn't this already have the correct shape? Why do you need to `H.expand(self._batch_shape)` below?